## Data Pre processing in Python
First, we'll be loading on the dataset.
The current dataset is the daily stock prices of Saudi Oil Company.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("Saudi Oil Company.csv")

## Printing the DataFrame
Notice, it has 250 rows and 7 columns

In [2]:
df

Date       Open       High        Low      Close  Adj Close  \
0    30/06/2021  32.000000  32.045452   31.81818  31.909090  31.044510   
1    01/07/2021  31.909090  31.909090   31.81818  31.818180  30.956062   
2    04/07/2021        NaN  31.954544   31.81818  31.863636  31.000288   
3    05/07/2021  31.818180  31.863636  31.727272  31.772726  30.911840   
4    06/07/2021  31.772726  31.818180  31.681818  31.772726  30.911840   
..          ...        ...        ...        ...        ...        ...   
245  26/06/2022  37.549999  37.849998  37.549999  37.650002  37.650002   
246  27/06/2022  37.799999  37.799999       37.5  37.599998  37.599998   
247  28/06/2022  37.650002  38.849998  37.650002  38.849998  38.849998   
248  29/06/2022  38.799999  39.599998       38.5  39.200001  39.200001   
249  30/06/2022  39.200001  39.200001         sd  38.900002  38.900002   

       Volume  
0    14452298  
1     5448521  
2     8335410  
3     5355944  
4     4323723  
..        ...  
245   2969194  
246   3064123  
247   5585162  
248   5682287  
249   5564245  

[250 rows x 7 columns]

## Info on the DF
Notice, how their are 2 columns that have non-null count less than others. This means that they have NAN values in them.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       250 non-null    object 
 1   Open       249 non-null    float64
 2   High       250 non-null    float64
 3   Low        250 non-null    object 
 4   Close      250 non-null    float64
 5   Adj Close  249 non-null    float64
 6   Volume     250 non-null    int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 13.8+ KB


Interesting thing to note is that the column low is of the type object even though it should be of the type float. 
</br> Sometimes, this is due to the column containing some non-float values. So, we'll check for it in the future

</br> Also, Date is an object, so we'll have to cast it to datatime object

### Quickly casting Date to datetime object

In [4]:
df['Date'] = pd.to_datetime(df['Date'])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       250 non-null    datetime64[ns]
 1   Open       249 non-null    float64       
 2   High       250 non-null    float64       
 3   Low        250 non-null    object        
 4   Close      250 non-null    float64       
 5   Adj Close  249 non-null    float64       
 6   Volume     250 non-null    int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 13.8+ KB


Notice how now the datatype is datetime!

## Casting Low from object to float
The errors = coerce, replaces non-numeric data with Nans

In [5]:
df['Low'] = pd.to_numeric(df['Low'],errors='coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       250 non-null    datetime64[ns]
 1   Open       249 non-null    float64       
 2   High       250 non-null    float64       
 3   Low        249 non-null    float64       
 4   Close      250 non-null    float64       
 5   Adj Close  249 non-null    float64       
 6   Volume     250 non-null    int64         
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 13.8 KB


Low's Datatype has changed from object to float but its non-null count has decreased. That means one of its entry was a non float value.

## Finding All NAN values
We can find all nan values, by using the command below. 
Here, isnull, checks if a value is NAN or not and the any method returns one value for each column. 

In [6]:
df[df.isnull().any(1)]

Date       Open       High        Low      Close  Adj Close   Volume
2   2021-04-07        NaN  31.954544  31.818180  31.863636  31.000288  8335410
38  2021-08-30  31.727272  31.863636  31.590908  31.863636        NaN  5126359
249 2022-06-30  39.200001  39.200001        NaN  38.900002  38.900002  5564245

## Handling NAN's
As, we had seen above, there are 3 nan values. And as this is way less than the total number of entries in the dataset (250) we can afford to simply remove the NAN values.
</br> This can be done by the code below

In [7]:
df = df.dropna()

df

Date       Open       High        Low      Close  Adj Close  \
0   2021-06-30  32.000000  32.045452  31.818180  31.909090  31.044510   
1   2021-01-07  31.909090  31.909090  31.818180  31.818180  30.956062   
3   2021-05-07  31.818180  31.863636  31.727272  31.772726  30.911840   
4   2021-06-07  31.772726  31.818180  31.681818  31.772726  30.911840   
5   2021-07-07  31.727272  31.818180  31.636362  31.772726  30.911840   
..         ...        ...        ...        ...        ...        ...   
244 2022-06-23  37.250000  37.599998  37.099998  37.200001  37.200001   
245 2022-06-26  37.549999  37.849998  37.549999  37.650002  37.650002   
246 2022-06-27  37.799999  37.799999  37.500000  37.599998  37.599998   
247 2022-06-28  37.650002  38.849998  37.650002  38.849998  38.849998   
248 2022-06-29  38.799999  39.599998  38.500000  39.200001  39.200001   

       Volume  
0    14452298  
1     5448521  
3     5355944  
4     4323723  
5     6504171  
..        ...  
244   4614276  
245   2969194  
246   3064123  
247   5585162  
248   5682287  

[247 rows x 7 columns]

Notice, how 3 rows have been dropped

## Ensuring Data conforms to ranges
We need to ensure that the various columns have </br>
 The value lies within the desired range

In [8]:
numeric_columns =  ["Open", "High", "Low", "Close", "Adj Close", "Volume"]
df = df[(df[numeric_columns] > 0).all(1)]
df

Date       Open       High        Low      Close  Adj Close  \
0   2021-06-30  32.000000  32.045452  31.818180  31.909090  31.044510   
1   2021-01-07  31.909090  31.909090  31.818180  31.818180  30.956062   
3   2021-05-07  31.818180  31.863636  31.727272  31.772726  30.911840   
4   2021-06-07  31.772726  31.818180  31.681818  31.772726  30.911840   
5   2021-07-07  31.727272  31.818180  31.636362  31.772726  30.911840   
..         ...        ...        ...        ...        ...        ...   
244 2022-06-23  37.250000  37.599998  37.099998  37.200001  37.200001   
245 2022-06-26  37.549999  37.849998  37.549999  37.650002  37.650002   
246 2022-06-27  37.799999  37.799999  37.500000  37.599998  37.599998   
247 2022-06-28  37.650002  38.849998  37.650002  38.849998  38.849998   
248 2022-06-29  38.799999  39.599998  38.500000  39.200001  39.200001   

       Volume  
0    14452298  
1     5448521  
3     5355944  
4     4323723  
5     6504171  
..        ...  
244   4614276  
245   2969194  
246   3064123  
247   5585162  
248   5682287  

[238 rows x 7 columns]

Notice, how the number of rows has reduced to 238. This means 9 rows had values less tha zero

## Describing the Dataframe

In [9]:
df.describe()

Open        High         Low       Close   Adj Close  \
count  238.000000  238.000000  238.000000  238.000000  238.000000   
mean    34.976852   35.243945   34.745263   35.010732   34.628942   
std      3.379243    3.480899    3.214491    3.345107    3.480300   
min     31.454544   31.590908   31.272726   31.363636   30.558058   
25%     31.920454   32.045452   31.863636   31.954544   31.560574   
50%     33.772728   34.000000   33.590908   33.795454   33.409122   
75%     38.487500   38.954544   37.920453   38.602273   38.413727   
max     42.500000   43.349998   42.200001   42.500000   42.155910   

             Volume  
count  2.380000e+02  
mean   8.385120e+06  
std    8.113957e+06  
min    1.447994e+06  
25%    4.043393e+06  
50%    5.560212e+06  
75%    8.829462e+06  
max    5.000824e+07

## Finding Correlation Matrix
We need to find a correlation matrix to see if we have some redundant columns

In [10]:
corr_matrix = df.corr()

corr_matrix

Open      High       Low     Close  Adj Close    Volume
Open       1.000000  0.995411  0.996604  0.992617   0.991396  0.348089
High       0.995411  1.000000  0.994215  0.997196   0.995963  0.402821
Low        0.996604  0.994215  1.000000  0.995599   0.994297  0.331673
Close      0.992617  0.997196  0.995599  1.000000   0.998456  0.378839
Adj Close  0.991396  0.995963  0.994297  0.998456   1.000000  0.367464
Volume     0.348089  0.402821  0.331673  0.378839   0.367464  1.000000

Notice, how open,high,low and close and adj close are all very highly correlated

## Removing redundant columns
<br> The following code makes sure to drop such columns.
<br> It first, gets the upper triangle of the correlational matrix (excluding the diagnal.)
<br> Then, we simply drop columns that have a high correlation with one other column. (Notice, how we won't drop both of the columns that have high correlation as we're only running on the upper triangle)

In [11]:
# Getting the upper triangle of the correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

# Find features with correlation greater than 0.95 and drop them.
to_drop = [column for column in upper.columns if any(upper[column] > 0.95)]
df.drop(to_drop, axis=1, inplace=True)
df.head()

C:\Users\SALMAN~1\AppData\Local\Temp/ipykernel_1264/1345639983.py:2: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))
c:\python39\lib\site-packages\pandas\core\frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


Date       Open    Volume
0 2021-06-30  32.000000  14452298
1 2021-01-07  31.909090   5448521
3 2021-05-07  31.818180   5355944
4 2021-06-07  31.772726   4323723
5 2021-07-07  31.727272   6504171

Finally, we can see that only 3 columns remain!

## The final step on this dataset, will be to normalize open and volume

In [12]:
#Updating numeric_cols after the drop of various columns
numeric_cols = ['Open', 'Volume']
df[numeric_cols]=(df[numeric_cols]-df[numeric_cols].min())/(df[numeric_cols].max()-df[numeric_cols].min())
df

c:\python39\lib\site-packages\pandas\core\frame.py:3191: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


Date      Open    Volume
0   2021-06-30  0.049383  0.267797
1   2021-01-07  0.041152  0.082383
3   2021-05-07  0.032922  0.080476
4   2021-06-07  0.028807  0.059220
5   2021-07-07  0.024691  0.104122
..         ...       ...       ...
244 2022-06-23  0.524691  0.065203
245 2022-06-26  0.551852  0.031326
246 2022-06-27  0.574486  0.033281
247 2022-06-28  0.560906  0.085197
248 2022-06-29  0.665021  0.087197

[238 rows x 3 columns]

Now, our data has been generically preprocessed though it still can be preprocessed more depending on how we're aiming to use it. However, for this particular example this is good enough!